src/threads.cpp

```diff
@@ -0,0 +1,328 @@
+  args[slen + alen] = 0;
+}
+
+int close(HANDLE h) { CloseHandle(h); }
+#endif
+
+#if defined(__unix) || defined(__unix__)
+
+Mutex::Mutex() { pthread_mutex_init(&m, NULL); }
+Mutex::~Mutex() { pthread_mutex_destroy(&m); }
+void Mutex::lock() {
+  //	printf("%p mutex.lock()\n", this);
+  pthread_mutex_lock(&m);
+}
+
+void Mutex::unlock() {
+  //	printf("%p mutex.unlock()\n", this);
+  pthread_mutex_unlock(&m);
+}
+
+#elif defined(_WIN32)
+
+Mutex::Mutex() { InitializeCriticalSection(&m); }
+Mutex::~Mutex() { DeleteCriticalSection(&m); }
+void Mutex::lock() { EnterCriticalSection(&m); }
+void Mutex::unlock() { LeaveCriticalSection(&m); }
+
+int WIN32_thread_create(HANDLE *tid, void *attr, void *(*func)(void *),
+                        void *arg) {
+  *tid = CreateThread(NULL, 0, (DWORD WINAPI(*)(void *))func, arg, 0, NULL);
+  if (!(*tid))
+    return 1;
+  else
+    return 0;
+};
+
+int WIN32_thread_join(HANDLE &tid, void **ret) {
+  DWORD tret;
+  if (WaitForSingleObject(tid, INFINITE) == WAIT_FAILED)
+    return -1;
+  if (ret) {
+    if (!GetExitCodeThread(tid, &tret))
+      return -1;
+    *ret = (void *)tret;
+  }
+  CloseHandle(tid);
+  tid = NULL;
+  return 0;
+};
+
+#endif
+
+// int createchild(char **argv, pipe_t &rdpipe, bool r, pipe_t &wrpipe, bool w)
+int createChildProcess(char **argv, pipe_t *rdpipe, pipe_t *wrpipe) {
+  printf("createchild(): pipes: %p, %p\n", rdpipe, wrpipe);
+
+#if defined(__unix) || defined(__unix__)
+  int cpid;
+
+  if (rdpipe) {
+    if (pipe(*rdpipe)) {
+      printf("Can't create pipe for stdout/stderr\n");
+      return -1;
+    }
+    printf("rdpipe = %d, %d\n", (*rdpipe)[0], (*rdpipe)[1]);
+  }
+
+  if (wrpipe) {
+    if (pipe(*wrpipe)) {
+      printf("Can't create pipe for stdout/stderr\n");
+      if (rdpipe) {
+        close(*rdpipe[0]);
+        close(*rdpipe[1]);
+      }
+      return -1;
+    }
+    printf("wrpipe = %d, %d\n", (*wrpipe)[0], (*wrpipe)[1]);
+  }
+
+  if ((cpid = fork()) == -1) {
+    printf("Can't fork()\n");
+    if (rdpipe) {
+      close((*rdpipe)[0]);
+      close((*rdpipe)[1]);
+    }
+    if (wrpipe) {
+      close((*wrpipe)[0]);
+      close((*wrpipe)[1]);
+    }
+    return -1;
+  }
+
+  if (!cpid) {
+    printf("child: fork() ok:)\n");
+    if (rdpipe) {
+      close((*rdpipe)[0]); // unused read end
+      close(STDOUT_FILENO);
+      close(STDERR_FILENO);
+      dup2((*rdpipe)[1], STDOUT_FILENO);
+      dup2((*rdpipe)[1], STDERR_FILENO);
+    }
+    if (wrpipe) {
+      close((*wrpipe)[1]); // unused write end
+      close(STDIN_FILENO);
+      dup2((*wrpipe)[0], STDIN_FILENO);
+    }
+    printf("starting %s...\n", argv[0]);
+    execvp(argv[0], argv);
+    printf("Can't exec() %s: %s\n", argv[0], strerror(errno));
+    _exit(0);
+  } else {
+    if (rdpipe)
+      close((*rdpipe)[1]); // unused write end
+    if (wrpipe)
+      close((*wrpipe)[0]); // unused read end
+  }
+  return cpid;
+
+#elif defined(_WIN32)
+  char args[1024] = "";
+  char *quoted_arg;
```

Comment:
Codacy found an issue: [The scope of the variable 'quoted_arg' can be reduced.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)